frontend/src/metabase/entities/containers/rtk-query/EntityListLoaderRtkQuery.tsx

```diff
@@ -0,0 +1,332 @@
+import type { ComponentType, ReactNode } from "react";
+import { useEffect, useMemo } from "react";
+import { useLatest } from "react-use";
+import { match } from "ts-pattern";
+
+import DefaultLoadingAndErrorWrapper from "metabase/components/LoadingAndErrorWrapper";
+import { capitalize } from "metabase/lib/formatting";
+import { useDispatch, useSelector } from "metabase/lib/redux";
+import {
+  setRequestError,
+  setRequestLoaded,
+  setRequestLoading,
+} from "metabase/redux/requests";
+import { isObject } from "metabase-types/guards";
+
+import type {
+  EntityDefinition,
+  EntityQuery,
+  EntityQuerySelector,
+  EntityType,
+  EntityTypeSelector,
+  ListMetadata,
+  ReloadInterval,
+  ReloadIntervalSelector,
+} from "./types";
+
+interface ChildrenProps<Entity, EntityWrapper> {
+  allError?: unknown;
+  allFetched?: boolean;
+  allLoaded?: boolean;
+  allLoading?: boolean;
+  entityQuery: EntityQuery;
+  error: unknown;
+  fetched: boolean;
+  loaded: boolean;
+  loading: boolean;
+  /**
+   * list item is EntityWrapper when Props["wrapped"] is true
+   */
+  list: EntityWrapper[] | Entity[] | undefined;
+  metadata: ListMetadata | undefined;
+  reload: () => void;
+}
+
+interface LoadingAndErrorWrapperProps {
+  children: ReactNode;
+  loading?: boolean;
+  error?: unknown;
+  noWrapper?: boolean;
+}
+
+interface Props<Entity, EntityWrapper> {
+  allError?: unknown;
+  allFetched?: boolean;
+  allLoaded?: boolean;
+  allLoading?: boolean;
+  ComposedComponent: (props: ChildrenProps<Entity, EntityWrapper>) => ReactNode;
+  entityQuery?: EntityQuery | EntityQuerySelector;
+  entityType: EntityType | EntityTypeSelector;
+  listName?: string;
+  loadingAndErrorWrapper?: boolean;
+  LoadingAndErrorWrapper?: ComponentType<LoadingAndErrorWrapperProps>;
+  reload?: boolean;
+  reloadInterval?: ReloadInterval | ReloadIntervalSelector<Entity>;
+  selectorName?: "getList" | "getListUnfiltered";
+  wrapped?: boolean;
+  onLoaded?: (list: Entity[]) => void;
+}
+
+const transformResponse = (fetched: unknown) => {
+  let results;
+  let metadata = {};
+
+  if (isObject(fetched) && fetched.data) {
+    const { data, ...rest } = fetched;
+    results = data;
+    metadata = rest;
+  } else {
+    results = fetched;
+  }
+
+  return { results, metadata };
+};
```

Comment:
Nit, I prefer using simple guards here.

```suggestion
const transformResponse = (fetched: unknown) => {
  if (!isObject(fetched) || !fetched.data) {
    return { data: fetched, metadata: {} }
  }   
  
  const { data, ...metadata } = fetched;
  return { data, metadata };
};
```

Reply:
329d2aaf25bdfc66260409de9047523f68aed949